### PR TITLE
initial bazelification of CDI

### PR DESCRIPTION
**Pull request by awels**
Signed-off-by: Alexander Wels <awels@redhat.com>

**What this PR does / why we need it**:
Update CDI build process to use bazel.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
CDI build now uses bazel.
```

### CODE REVIEW

File: hack/build/bazel-docker.sh [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+# append .docker directory as volume
+mkdir -p "${HOME}/.docker"
+volumes="$volumes -v ${HOME}/.docker:/root/.docker:ro,z"
+
+# Ensure that a bazel server is running
+
+if [ -z "$(docker ps --format '{{.Names}}' | grep ${BUILDER_VOLUME}-bazel-server)" ]; then
+	echo "Starting bazel-server"
+	echo "docker run --network host -d ${volumes} --security-opt label:disable --name ${BUILDER_VOLUME}-bazel-server -w \"/root/go/src/kubevirt.io/containerized-data-importer\" --rm ${BUILDER_TAG} hack/build/bazel-server.sh"
+    docker run --network host -d ${volumes} --security-opt label:disable --name ${BUILDER_VOLUME}-bazel-server -w "/root/go/src/kubevirt.io/containerized-data-importer" --rm ${BUILDER_TAG} hack/build/bazel-server.sh
```
Comment (mhenriks) [inline hack/build/bazel-docker.sh]: maybe assign a name to this container so it's easier to manually kill later?

File: hack/build/bazel-docker.sh [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+volumes="-v ${BUILDER_VOLUME}:/root:rw,z"
+
+# append .docker directory as volume
+mkdir -p "${HOME}/.docker"
+volumes="$volumes -v ${HOME}/.docker:/root/.docker:ro,z"
+
+# Ensure that a bazel server is running
+
+if [ -z "$(docker ps --format '{{.Names}}' | grep ${BUILDER_VOLUME}-bazel-server)" ]; then
```
Comment (mhenriks) [inline hack/build/bazel-docker.sh]: maybe have a variable equal to `${BUILDER_VOLUME}-bazel-server`?

File: Makefile [hunk 3]
```diff
@@ -12,32 +12,33 @@
...
-		cluster-up cluster-down cluster-sync cluster-sync-controller cluster-sync-cloner cluster-sync-importer cluster-sync-apiserver cluster-sync-uploadproxy cluster-sync-uploadserver \
+		cluster-up cluster-down cluster-sync \
 		test test-functional test-unit test-lint \
 		publish \
 		vet \
 		format \
 		goveralls \
-		release-description
+		release-description \
+		bazel-generate bazel-build bazel-build-images bazel-push-images
```
Comment (mhenriks) [inline Makefile]: not sure `bazel-push-images` should be here.  We didn't push images before when simply running `make`.  This basically requires `make cluster-sync` to have been run first.

**Comment by mhenriks** [comment]
What is the deal with these links created to /root
```
lrwxrwxrwx.  1 mhenriks mhenriks      113 Oct  6 17:57 bazel-testlogs -> /root/.cache/bazel/_bazel_root/df081773a313729c04cf85e830afc1f9/execroot/__main__/bazel-out/k8-fastbuild/testlogs
lrwxrwxrwx.  1 mhenriks mhenriks       91 Oct  6 17:57 bazel-out -> /root/.cache/bazel/_bazel_root/df081773a313729c04cf85e830afc1f9/execroot/__main__/bazel-out
lrwxrwxrwx.  1 mhenriks mhenriks      108 Oct  6 17:57 bazel-genfiles -> /root/.cache/bazel/_bazel_root/df081773a313729c04cf85e830afc1f9/execroot/__main__/bazel-out/k8-fastbuild/bin
lrwxrwxrwx.  1 mhenriks mhenriks       81 Oct  6 17:57 bazel-containerized-data-importer -> /root/.cache/bazel/_bazel_root/df081773a313729c04cf85e830afc1f9/execroot/__main__
lrwxrwxrwx.  1 mhenriks mhenriks      108 Oct  6 17:57 bazel-bin -> /root/.cache/bazel/_bazel_root/df081773a313729c04cf85e830afc1f9/execroot/__main__/bazel-out/k8-fastbuild/bin
```

**Comment by awels** [comment]
There is still an issue in make bazel-generate, it is supposed to rsync which will delete those files. But it dies before it gets there.

File: Makefile (hunk 3 above)
Comment (awels) [reply Makefile]: bazel-push-images builds the images and pushes it to the configured registry. make cluster-sync does that and install CDI as well.

Comment (mhenriks) [reply Makefile]: Sorry, I meant `make cluster-up`.  Basically, I don't think we should always push to the registry.  Sometimes I just want to make sure everything compiles/builds locally.

Comment (awels) [reply Makefile]: So those are just entries in the .PHONY section of the makefile, telling it there is no requirement that a file called bazel-generate/etc exist. I am not sure what you mean by make sure everything compiles/builds locally. That is what make bazel-build is for, that builds the binaries.

File: Makefile [hunk 4]
```diff
@@ -12,32 +12,33 @@
...
 ifeq (${DOCKER}, 1)
 DO=./hack/build/in-docker.sh
+DO_BAZ=./hack/build/bazel-docker.sh
 else
 DO=eval
+DO_BAZ=eval
 endif
 
-all: manifests docker
+all: manifests push
```
Comment (mhenriks) [inline Makefile]: I guess the issue I have is the 'push' here.  we didn't do that before.

**Comment by awels** [comment]
ci test please

File: hack/build/docker/builder/entrypoint.sh [hunk 5]
```diff
@@ -25,4 +25,4 @@ eval "$@"
 
 if [ -n ${RUN_UID} ] && [ -n ${RUN_GID} ]; then
     find . -user root -exec chown -h ${RUN_UID}:${RUN_GID} {} \;
-fi
+fi
```
Comment (mhenriks) [inline hack/build/docker/builder/entrypoint.sh]: No newline

Comment (awels) [reply hack/build/docker/builder/entrypoint.sh]: Fixed

**Comment by awels** [comment]
Perfect, thanks @rmohr

File: hack/build/bazel-docker.sh (hunk 2 above)
Comment (awels) [reply hack/build/bazel-docker.sh]: Done

**Comment by rmohr** [comment]
/hold

Ups an error

**Comment by rmohr** [comment]
The URLs are misspelled

**Comment by awels** [comment]
rebased on new kubevirtci image

**Comment by awels** [comment]
ci test please

**Comment by awels** [comment]
ci test please

**Pull request merged**